### DISCUSSION TITLE
Using CROCO with FABM - work involved?

### DISCUSSION THREAD

**Discussion by jvmcgovern** [fabm-model/fabm#42] [Q&A] [answered]
Hi there,

I'd like to know what level of work required to get FABM to handle CROCO for hydrodynamics? It is an offshoot of ROMS ("French ROMS" and an evolution of "ROMS-AGRIF") and I expect it would be analogous to set up but would like to make a firm enquiry and do what is possible to make it happen.

All the best,
Joe

**Comment by bolding** [comment, marked as answer]
Hi Joe

Jorn - who is best suited for answering this - is on vacation for yet some time so I'll provide a correct but not very elaborate answer.

As there already is a coupling with ROMS and if CROCO does not deviate to much it is all possible. In principle the coupling is rather simple and the hydrodynamic model takes care of the most difficult parts. The wiki has quite extensive documentation on the steps to couple to 1D and 3D models. One issue with the ROMS coupling is that ROMS requires the number of bio-geochemical state variables to be known at compile time - which is against the philosophy of FABM where everything is configurable at run time. To overcome I think they use a small utility that parses the fabm.yaml file and reports the size needed for compilation.

So in short - the CROCO build system must be modified to link to the FABM libraries and a coupling source code file must be created with the necessary FABM calls to initialize, update and finalize and these calls must be integrated in CROCO. 

Best regards,
Karsten

**Reply by jvmcgovern** [reply]
Hi Karsten,

Thank you for explaining the situation. 

I would like to discuss this further with CROCO developers later this year as it may be of interest to them. If I get a favourable response I will revert to here to continue the conversation. 

Best regards,
Joe

**Reply by pwallhead** [reply]
Hi Joe

As Karsten says, it depends how much CROCO deviates from the other ROMS versions that have been coupled already (ROMS trunk, COAWST, Hedstrom branch).  I don't know the CROCO code, but I would be surprised if there were major differences that affect the FABM coupling.  I suspect it's mostly a case of copy-pasting code segments from the existing FABM-adapted files into the corresponding CROCO files  (I would probably start with the trunk version file set https://github.com/pwallhead/roms_fabm/tree/main/ROMS_ARANGO_FABM/Src_ARANGO_20012020_modify/, see also ROMS_ARANGO_FABM/roms_arango_fabm_notes.txt).  As Karsten also mentioned, since ROMS requires the configuration to be fixed at compilation we have a python script that reads the fabm.yaml file and generates the needed ROMS input files for compilation (see roms_fabm/README.txt).

Best
Phil

**Reply by jvmcgovern** [reply]
Hi @pwallhead @bolding @jornbr @platipodium 

There is now interest in this from the CROCO community and from SINTEF Ocean, see link below. 

https://forum.croco-ocean.org/t/croco-and-fabm/470

Would you be free to meet next week via teams to discuss? Wednesday or Friday would suit best if possible.

Best regards,
Joe

**Comment by jornbr** [comment]
Hi @jvmcgovern,

That sounds great! Happy to hear what you're planning and to think along. I'll follow up via email.

Best,

Jorn

**Comment by bolding** [comment]
Hello

The times given by Jorn works well for me as well.

Karsten
